### ISSUE TITLE
Contextual return types for expression-bodied members

### ISSUE THREAD

**Issue by SLaks** [dotnet/roslyn#16044]
```C#
class C {
	partial A => new partial();
	async B => new async();
}

class partial {}
class async {}
```

The first line compiles; the second one does not.

Why not?
Why is `async` different from `partial`?

http://tryroslyn.azurewebsites.net/#b:master/f:>ilr/MYGwhgzhAEDC0G8BQBIADmATgFwJZhGgEFoBeAPmgDsBTAd2gx3xAAoBKAblUgE8rg0AEJlKtBnwEduAXyRJQkGEzwFEcxVGiTBCGUA=

**Comment by DavidArno** [comment]
`async` is in scope as a keyword in this example as it could be the start of an `async B f() { ...` method declaration.

**Comment by alrz** [comment]
@DavidArno So is `partial` e.g partial methods. Note that it will compile in there cases
```cs
async M() => new async();
async M { get => new async(); }
```
etc. This is probably not handled correctly in the parser.

**Comment by DavidArno** [comment]
Oh, I forgot about partial methods. In that case it does indeed appear to be a parser inconsistency issue.

**Comment by gafter** [comment]
`partial` is a contextual keyword. The context is that the following token must be `class`, `struct`, or `void`.

`async` is a contextual keyword. It is treated as a keyword when it appears in a position that could be a member modifier.

They are different because they are specified to be different.

**Comment by gafter** [comment]
The spec does say, however

> contextual keywords can conflict with declared names. In such cases, the declared name takes precedence over the use of the identifier as a contextual keyword.

**Comment by jcouv** [comment]
Fix is in for 15.7 preview 4.

**Comment by jcouv** [comment]
Thanks @alrz for the fix!
